### PR TITLE
DEV-6954 - Adds config to make the ES Scheduler work with IP per container executors

### PR DESCRIPTION
**Pull request by souza-dan**
https://datatamr.atlassian.net/browse/DEV-6954

Upgrades the mesos library version and adds a number of flags to support the IP per container.

`"--mesosOfferIgnorePorts <false|true>"`
Toggle ports from being considered in offers. With IP per container, we no longer care about managing ports with mesos since every container has its own IP and set of ports, so this can be set to true.

`"--mesosTaskDockerNetwork <host|user|bridge|none>"`
Specify the type of docker network. This was hard coded to HOST, and needs to be settable to USER to enable IP per container.

`"--mesosTaskNetworkInfoName <network name>"`
Use when the docker network is "user". Need to specify the name of the network, which should be the same as the framework name (but could maybe be anything).

`"--mesosMultipleTasksPerHost <false|true>"`
Toggle the offer check that limits only one executor per mesos agent. We want to support an arbitrary number of executors not bounded by the cluster's number of agents. True allows multiple executors per mesos agent.

`"--mesosOfferWaitForRunning <false|true>"`
Toggle the offer check that waits for one executor to start. With IP per container enabled, the scheduler has a reference to the IP of the agent the container is running on, but not the container IP so it can't check if it's running yet. It does this check so that it can give subsequent executors the first executor's IP to do master discovery/election. We're bypassing this requirement by specifying the host it should use for explicitly with `--discoveryZenPingUnicastHosts`.

`"--discoveryZenPingUnicastHosts <comma separated list of es executors>"`
Provide a host of an executor to do master election. Scheduler doesn't have hooks to get the IP of the container (just the agent it's running on), so we can specify the executor's mesos DNS hostname which looks something like: `elasticsearch-executor.elasticsearch-${FRAMEWORK_NAME}.${MESOS_DNS_TLD}`.

### CODE REVIEW

**Comment by laferrieren** [comment]
Not the best with java, but lgtm.  Might want to have someone with java experience take a look.

File: scheduler/src/main/java/org/apache/mesos/elasticsearch/scheduler/OfferStrategyExternalStorage.java [hunk 1]
```diff
@@ -11,19 +11,21 @@
...
-                new OfferRule("Cluster size already fulfilled", offer -> clusterState.getTaskList().size() >= configuration.getElasticsearchNodes()),
-                new OfferRule("Offer did not have 2 ports", offer -> !containsTwoPorts(offer.getResourcesList())),
-                new OfferRule("The offer does not contain the user specified ports", offer -> !containsUserSpecifiedPorts(offer.getResourcesList())),
-                new OfferRule("Offer did not have enough CPU resources", offer -> !isEnoughCPU(configuration, offer.getResourcesList())),
-                new OfferRule("Offer did not have enough RAM resources", offer -> !isEnoughRAM(configuration, offer.getResourcesList()))
+          new OfferRule("Host already running task", offer -> !configuration.getMesosMultipleTasksPerHost() && isHostAlreadyRunningTask(offer)),
+          new OfferRule("Hostname is unresolveable", offer -> !isHostnameResolveable(offer.getHostname())),
+          new OfferRule("First ES node is not responding", offer -> !(isAtLeastOneESNodeRunning() || !configuration.getMesosOfferWaitForRunning())),
```
Comment (okrivche) [inline scheduler/src/main/java/org/apache/mesos/elasticsearch/scheduler/OfferStrategyExternalStorage.java]: May make it slightly easier to read if this negation was distributed out.

File: scheduler/src/main/java/org/apache/mesos/elasticsearch/scheduler/OfferStrategyNormal.java [hunk 2]
```diff
@@ -14,17 +14,18 @@
...
-                new OfferRule("Offer did not have 2 ports", offer -> !containsTwoPorts(offer.getResourcesList())),
-                new OfferRule("The offer does not contain the user specified ports", offer -> !containsUserSpecifiedPorts(offer.getResourcesList())),
-                new OfferRule("Offer did not have enough CPU resources", offer -> !isEnoughCPU(configuration, offer.getResourcesList())),
-                new OfferRule("Offer did not have enough RAM resources", offer -> !isEnoughRAM(configuration, offer.getResourcesList())),
-                new OfferRule("Offer did not have enough disk resources", offer -> !isEnoughDisk(configuration, offer.getResourcesList()))
+          new OfferRule("Host already running task", offer -> !configuration.getMesosMultipleTasksPerHost() && isHostAlreadyRunningTask(offer)),
+          new OfferRule("Hostname is unresolveable", offer -> !isHostnameResolveable(offer.getHostname())),
+          new OfferRule("First ES node is not responding", offer -> !(isAtLeastOneESNodeRunning() || !configuration.getMesosOfferWaitForRunning())),
```
Comment (okrivche) [inline scheduler/src/main/java/org/apache/mesos/elasticsearch/scheduler/OfferStrategyNormal.java]: Here too.

File: scheduler/src/test/java/org/apache/mesos/elasticsearch/scheduler/ConfigurationTest.java [hunk 3]
```diff
@@ -63,19 +64,148 @@ public void shouldCreateArguments() {
...
 
+    @Test
+    public void shouldCreateArgumentsWithHostname() {
+        final String discoveryHostname = "ahostnameidefined";
+        Configuration configuration = new Configuration(ZookeeperCLIParameter.ZOOKEEPER_MESOS_URL, "aa", Configuration.DISCOVERY_ZEN_PING_UNICAST_HOSTS, discoveryHostname);
+        final ClusterState clusterState = Mockito.mock(ClusterState.class);
+        final int port = 1234;
+        final Protos.DiscoveryInfo discoveryInfo = Protos.DiscoveryInfo.newBuilder().setPorts(Protos.Ports.newBuilder()
+          .addPorts(Protos.Port.newBuilder().setNumber(port))
+          .addPorts(Protos.Port.newBuilder().setNumber(port)))
```
Comment (okrivche) [inline scheduler/src/test/java/org/apache/mesos/elasticsearch/scheduler/ConfigurationTest.java:85]: Nit: I'm not familiar with DiscoveryInfo... but do you need to add multiple ports here?

File: scheduler/src/test/java/org/apache/mesos/elasticsearch/scheduler/ConfigurationTest.java [hunk 4]
```diff
@@ -63,19 +64,148 @@ public void shouldCreateArguments() {
...
+        Configuration configuration = new Configuration(ZookeeperCLIParameter.ZOOKEEPER_MESOS_URL, "aa", Configuration.MESOS_TASK_DOCKER_NETWORK, "NONE");
+        assertEquals(configuration.getTaskDockerNetworkProtos(), Protos.ContainerInfo.DockerInfo.Network.NONE);
+    }
+
+    @Test
+    public void shouldSetMesosDockerNetworkDefault() throws UnknownHostException {
+        Configuration configuration = new Configuration(ZookeeperCLIParameter.ZOOKEEPER_MESOS_URL, "aa");
+        assertEquals(configuration.getTaskDockerNetworkProtos(), Protos.ContainerInfo.DockerInfo.Network.HOST);
+
+        configuration = new Configuration(ZookeeperCLIParameter.ZOOKEEPER_MESOS_URL, "aa", Configuration.MESOS_TASK_DOCKER_NETWORK, "notreal");
```
Comment (okrivche) [inline scheduler/src/test/java/org/apache/mesos/elasticsearch/scheduler/ConfigurationTest.java]: This seems to be the same test as `shouldSetMesosDockerNetworkBadInput()`

**Review by okrivche — COMMENTED** [review]
Just a few small comments; also, a .DS_Store file snuck in. Otherwise looks reasonable.

File: scheduler/src/test/java/org/apache/mesos/elasticsearch/scheduler/ConfigurationTest.java (hunk 3 above)
Comment (souza-dan) [reply scheduler/src/test/java/org/apache/mesos/elasticsearch/scheduler/ConfigurationTest.java:85]: Oddly enough it does (as strange as this looks). I thought the same thing when I wrote the test.

Needs ports for client and transport, and uses indices set in `Discovery` class to grab them out of `DiscoveryInfo`. Test ends in a NPE if you don't set two ports here.
```   public class Discovery {
    // An elasticsearch task has 2 ports, client port and transport port.
    public static final int CLIENT_PORT_INDEX = 0;
    public static final int TRANSPORT_PORT_INDEX = 1;
    public static final String CLIENT_PORT_NAME = "CLIENT_PORT";
    public static final String TRANSPORT_PORT_NAME = "TRANSPORT_PORT";
    public static final int EXPECTED_NUMBER_OF_PORTS = 2;
}
```

**Pull request merged**